### DISCUSSION TITLE
Version Control Considerations and Workflows in a Site Editing paradigm

### DISCUSSION THREAD

**Discussion by colorful-tones** [WordPress/developer-blog-content#249] [Topic Ideas]
I've heard this issue come up a few times from several different developers:

> It’s very hard to find anywhere a progress and a talk about FSE version control. I’m not talking about revisions, as this is database dependent - but rather how could we keep the functionality of the project synced via git.

I would like to open this up for discussion and the possibility of somebody promoting this to a formal topic and finding publication.

**Comment by colorful-tones** [comment]
@bacoords shared this very relevant and related Gutenberg Discussion: https://github.com/WordPress/gutenberg/discussions/59480

**Comment by colorful-tones** [comment]
Also, it was discussed in a recent Hallway Hangout: https://www.youtube.com/watch?v=KjQwA_Uz_kQ

**Comment by justintadlock** [comment]
For this to become a topic, someone would likely need to propose one or more solutions and share how to accomplish the desired goal. Right now, the proposal is just "here's a problem."

I have a solution that I feel like works for my case on the more product/distributed side of things, but the real issue with version control and FSE is with agency work. I haven't heard a lot of good solutions when talking about maintaining changes between the database and the filesystem.

**Reply by bacoords** [reply]
There's still a few gaps that need to be solved before we can say "here's a good approach", though Create Block Theme (CBT) is helping. 

Off the top of my head
- Partially-Synced Patterns are going to be critical for controlling design across a site
- Two-way syncing of Patterns between the theme & database (similar to how CBT can export templates to your theme and 'reset' them in the DB). There's a plugin out there called Pattern Editor that's made a decent start here, but it should really be in CBT.
- Navigation Block ability to use menu "locations" rather than post IDs so that template parts like headers and footers can be controlled via a theme, specifically if you're using a theme across multiple sites.
- Better permissions controls so that nothing is changed in the site editor on production (or if it is, there needs to be some way to get those changes back into your repository, which sounds harder)

I have faith that there'll be a great workflow in a year from now, but there's still work to be done.

**Comment by colorful-tones** [comment]
> For this to become a topic, someone would likely need to propose one or more solutions and share how to accomplish the desired goal. Right now, the proposal is just "here's a problem."

Agree. Thanks for clarifying here @justintadlock 👍 

Also, I just wanted to share a tiny piece of this puzzle that has one solution (in case somebody stumbles across here and is faced with the version control challenge): some of the pages on WordPress.org run off the block editor and have a version control flow. From @ryelle 

> we don’t allow site editor access, the templates are (basically) manually edited. but the post editor for each page controls everything on the page except the global header & footer. That’s how we can have the homepage, [release page](https://wordpress.org/download/releases/6-5/), and more [general text pages](https://wordpress.org/about/accessibility/) all managed in the same way. it’s definitely not the process most devs would take - we also need to run some processing to wrap strings in translation functions - but the idea could be expanded to wp_template & wp_template_part data.

**Comment by bph** [comment]
There is a similar topic suggested: [Creating a low-code block theme development workflow with WordPress Playground and the Create Block Theme plugin](https://github.com/WordPress/developer-blog-content/discussions/266)
Also, we scheduled a Hallway Hangout to discuss such an approach for June 19 at 11:00 UTC

**Comment by bph** [comment]
On the Make Blog:  [Recap Hallway Hangout: Theme Building with Playground, Create-block-theme plugin, and GitHub](https://make.wordpress.org/core/2024/06/25/recap-hallway-hangout-theme-building-with-playground-create-block-theme-plugin-and-github/)

**Comment by bph** [comment]
With the published post [Streamlining block theme development with WordPress Playground and GitHub](https://developer.wordpress.org/news/2026/01/streamlining-block-theme-development-with-wordpress-playground-and-github/) this topic has been covered for now.
